biom/table.py

```diff
@@ -4737,6 +4737,95 @@ def to_json(self, generated_by, direct_io=None):
                                       matrix_element_type, shape,
                                       u''.join(data), rows, columns])
 
+    @staticmethod
+    def from_adjacency(lines):
+        """Parse an adjacency format into BIOM
+
+        Parameters
+        ----------
+        lines : list, str, or file-like object
+            The tab delimited data to parse
+
+        Returns
+        -------
+        biom.Table
+            A BIOM ``Table`` object
+
+        Notes
+        -----
+        The input is expected to be of the form: observation, sample, value. A
+        header is not required, but if present, it must be of the form:
+
+        #OTU ID<tab>SampleID<tab>value
+
+        Examples
+        --------
+        Parse tab separated adjacency data into a table:
+
+        >>> from biom.table import Table
+        >>> from io import StringIO
+        >>> data = 'a\\tb\\t1\\na\\tc\\t2\\nd\\tc\\t3'
+        >>> data_fh = StringIO(data)
+        >>> func = lambda x : x
+        >>> test_table = Table.from_adjacency(data_fh)
+        """
+        if not isinstance(lines, (list, tuple)):
+            if hasattr(lines, 'readlines'):
+                lines = lines.readlines()
+            elif hasattr(lines, 'splitlines'):
+                lines = lines.splitlines()
+            else:
+                raise ValueError("Not sure how to handle this input")
+
+        def is_num(item):
+            # from https://stackoverflow.com/a/23059703
+            numeric = re.compile(r'(?=.)([+-]?([0-9]*)(\.([0-9]+))?)([eE][+-]?\d+)?')  # noqa
+            match = numeric.match(item)
+            start, stop = match.span()
+            if (stop - start) == len(item):
+                return True
+            else:
+                return False
+
+        # sanity check and determine if we have a header or not
+        lh = lines[0].strip().split('\t')
+        if len(lh) != 3:
+            raise ValueError("Does not appear to be an adjacency format")
+        elif lh == ['#OTU ID', 'SampleID', 'value']:
+            include_line_zero = False
+        elif not is_num(lh[0]) and not is_num(lh[1]) and is_num(lh[2]):
```

Comment:
From what I gather here only values are allowed to be numbers? Aren't there cases where OTU ids or sample ids could look like numbers but really be identifiers instead? I think this would trigger an error for a valid line, right?

Reply:
Fair. So maybe allow cols 1 and 2 to be anything, and only assert column 3 is numeric?